extensions/data-transfer/portability-data-transfer-smugmug/src/main/java/org/dataportabilityproject/transfer/smugmug/photos/SmugMugInterface.java

```diff
@@ -70,22 +88,88 @@
+    this.user = getUserInformation().getUser();
   }
 
-  SmugMugResponse<SmugMugAlbumInfoResponse> makeAlbumInfoRequest(String url) throws IOException {
-    return makeRequest(url, new TypeReference<SmugMugResponse<SmugMugAlbumInfoResponse>>() {});
+  /* Returns the album information corresponding to the album URI provided. */
+  SmugMugAlbumInfoResponse getAlbumInfo(String url) throws IOException {
+    Preconditions.checkArgument(
+        !Strings.isNullOrEmpty(url), "Album URI is required to retrieve album information");
+    return makeRequest(url, new TypeReference<SmugMugResponse<SmugMugAlbumInfoResponse>>() {})
+        .getResponse();
   }
 
-  SmugMugResponse<SmugMugAlbumsResponse> makeAlbumRequest(String url) throws IOException {
-    return makeRequest(url, new TypeReference<SmugMugResponse<SmugMugAlbumsResponse>>() {});
+  /* Returns the album corresponding to the url provided. If the url is null or empty, this
+   * returns the top level user albums. */
+  SmugMugAlbumsResponse getAlbums(String url) throws IOException {
+    if (Strings.isNullOrEmpty(url)) {
+      url = user.getUris().get(ALBUMS_KEY).getUri();
+    }
+    return makeRequest(url, new TypeReference<SmugMugResponse<SmugMugAlbumsResponse>>() {})
+        .getResponse();
   }
 
-  SmugMugResponse<SmugMugUserResponse> makeUserRequest() throws IOException {
-    return makeRequest(USER_URL, new TypeReference<SmugMugResponse<SmugMugUserResponse>>() {});
+  /* Creates an album with albumName provided. */
+  SmugMugAlbumResponse createAlbum(String albumName) throws IOException {
+    // Set up album
+    Map<String, String> json = new HashMap<>();
+    String niceName = "Copy-of-" + albumName.replace(' ', '-');
+    json.put("UrlName", niceName);
+    // Allow conflicting names to be changed
+    json.put("AutoRename", "true");
+    json.put("Name", "Copy of " + albumName);
+    // All imported content is private by default.
+    json.put("Privacy", "Private");
+    HttpContent content = new JsonHttpContent(new JacksonFactory(), json);
+
+    // Upload album
+    String folder = user.getUris().get(FOLDER_KEY).getUri();
+
+    SmugMugResponse<SmugMugAlbumResponse> response =
+        postRequest(
+            folder + "!albums",
+            content,
+            ImmutableMap.of(),
+            new TypeReference<SmugMugResponse<SmugMugAlbumResponse>>() {});
+
+    checkState(response.getResponse() != null, "Response is null");
+    checkState(response.getResponse().getAlbum() != null, "Album is null");
+
+    return response.getResponse();
   }
 
-  <T> SmugMugResponse<T> makeRequest(String url, TypeReference<SmugMugResponse<T>> typeReference)
-      throws IOException {
+  /* Uploads the resource at photoUrl to the albumId provided
+   * The albumId must exist before calling upload, else the request will fail */
+  ImageUploadResponse uploadImage(String photoUrl, String albumId) throws IOException {
```

Comment:
uploadImage is specific to the importer only, right? I think it should belong in the importer then

Reply:
It is, but since it does a lot of specific  smugmug http header and body modifications, I think it made more sense to leave it with the smugmug interface (because also that is what will be signing the post request).